File: src/goal/managedTodoState.ts
/**
 * Managed-goal todo state contract.
 *
 * This module owns the JSON todo mapping. State repo transport stays in
 * stateStore.ts.
 */
import { type GoalState, parseGoalState, serializeGoalState } from "./state.js"

export interface TodoItemState {
  id: string
  title: string
  body: string
  assignee: string | null
  completed: boolean
  createdAt: string
  completedAt: string | null
  meta?: Record<string, unknown>
}

export function parseTodoGoalState(goalId: string, filePath: string, raw: string): GoalState {
  const data = parseJsonRecord(raw) ?? {}
  const items = normalizeItems(data.items)
  const destination = recordField(data.destination)
  const evidence =
    stringArray(destination.evidence).length > 0
      ? stringArray(destination.evidence)
      : stringArray(data.evidence).length > 0
        ? stringArray(data.evidence)
        : items.map((item) => stringField(recordField(item.meta).evidence) || item.id).filter(Boolean)
  const facts = {
    ...recordField(data.facts),
    ...Object.fromEntries(
      items.map((item) => [stringField(recordField(item.meta).evidence) || item.id, item.completed]),
    ),
  }
  const route = Array.isArray(data.route) ? (data.route as Record<string, unknown>[]) : routeFromItems(items)

  return parseGoalState(filePath, {
    ...data,
    id: goalId,
    version: data.version ?? 1,
    state: data.state ?? "active",
    type: data.type ?? "general",
    destination: {
      ...destination,
      outcome: stringField(data.description) || stringField(destination.outcome),
      evidence,
    },
    capabilities:
      stringArray(data.capabilities).length > 0
        ? stringArray(data.capabilities)
        : route.map((step) => stringField(step.capability)).filter(Boolean),
    route,
    facts,
    blockers: stringArray(data.blockers),
  })
}

export function isManagedTodoRaw(raw: string): boolean {
  return isManagedTodoRecord(parseJsonRecord(raw) ?? {})
}

export function serializeTodoGoalState(goalId: string, state: GoalState, previousRaw?: string): string {
  const raw = JSON.parse(serializeGoalState(state)) as Record<string, unknown>
  const destination = recordField(raw.destination)
  const outcome = stringField(raw.description) || stringField(destination.outcome)
  const evidence =
    stringArray(destination.evidence).length > 0 ? stringArray(destination.evidence) : stringArray(raw.evidence)
  const route = Array.isArray(raw.route) ? (raw.route as Record<string, unknown>[]) : []
  const facts = recordField(raw.facts)
  const evidenceState = recordField(raw.evidenceState)
  const now = new Date().toISOString()
  const createdAt = stringField(raw.createdAt) || stringField(raw.startedAt) || now
  const routeByEvidence = new Map(route.map((step) => [stringField(step.evidence), step] as const))
  const previousItems = new Map(parseItemsFromAnyRaw(previousRaw ?? "").map((item) => [item.id, item] as const))
  const items =
    evidence.length > 0
      ? evidence.map((key) =>
          itemFromEvidence(key, routeByEvidence.get(key), facts, evidenceState, createdAt, now, previousItems.get(key)),
        )
      : stringArray(raw.capabilities).map((capability) =>
          itemFromCapability(capability, createdAt, previousItems.get(capability)),
        )

  return `${JSON.stringify(
    {
      version: 1,
      ...raw,
      id: goalId,
      title: goalId,
      description: outcome,
      createdAt,
      managed: true,
      managedModel: raw.scheduleMode === "agentLoop" || raw.type === "agentLoop" ? "agentLoop" : "agentGoal",
      evidence,
      items,
    },
    null,
    2,
  )}\n`
}

function isManagedTodoRecord(record: Record<string, unknown>): boolean {
  return (
    record.managed === true ||
    record.managed === "true" ||
    record.managedModel === "agentGoal" ||
    record.managedModel === "agentLoop"
  )
}

function itemFromEvidence(
  evidence: string,
  step: Record<string, unknown> | undefined,
  facts: Record<string, unknown>,
  evidenceState: Record<string, unknown>,
  createdAt: string,
  now: string,
  prior?: TodoItemState,
): TodoItemState {
  const completed = facts[evidence] === true
  const progress = recordField(evidenceState[evidence])
  return {
    id: evidence,
    title: (prior?.title ?? stringField(step?.stage)) || evidence,
    body: prior?.body ?? "",
    assignee: prior?.assignee ?? null,
    completed,
    createdAt: prior?.createdAt ?? createdAt,
    completedAt: completed ? (prior?.completedAt ?? now) : null,
    meta: {
      ...(prior?.meta ?? {}),
      evidence,
      ...(stringField(progress?.resultClass) ? { resultClass: stringField(progress?.resultClass) } : {}),
      ...(typeof progress?.attempts === "number" ? { attempts: progress.attempts } : {}),
      ...(stringField(progress?.reason) ? { reason: stringField(progress?.reason) } : {}),
      ...(stringField(progress?.nextAction) ? { nextAction: stringField(progress?.nextAction) } : {}),
      ...(stringField(progress?.nextRetryAt) ? { nextRetryAt: stringField(progress?.nextRetryAt) } : {}),
      ...(typeof progress?.issue === "number" ? { issue: progress.issue } : {}),
      ...(step
        ? {
            stage: stringField(step.stage),
            capability: stringField(step.capability),
            ...(step.args && typeof step.args === "object" ? { args: step.args } : {}),
            ...(step.saveReport === true ? { saveReport: true } : {}),
            ...(step.onPending && typeof step.onPending === "object" ? { onPending: step.onPending } : {}),
            ...(step.onFailure && typeof step.onFailure === "object" ? { onFailure: step.onFailure } : {}),
          }
        : {}),
    },
  }
}

function itemFromCapability(capability: string, createdAt: string, prior?: TodoItemState): TodoItemState {
  return {
    id: capability,
    title: prior?.title ?? capability,
    body: prior?.body ?? "",
    assignee: prior?.assignee ?? null,
    completed: prior?.completed ?? false,
    createdAt: prior?.createdAt ?? createdAt,
    completedAt: prior?.completedAt ?? null,
    meta: { ...(prior?.meta ?? {}), capability },
  }
}

function routeFromItems(items: TodoItemState[]): Record<string, unknown>[] {
  return items.flatMap((item) => {
    const meta = recordField(item.meta)
    const evidence = stringField(meta.evidence) || item.id
    const stage = stringField(meta.stage)
    const capability = stringField(meta.capability)
    if (!evidence || !stage || !capability) return []
    return [
      {
        evidence,
        stage,
        capability,
        ...(meta.args ? { args: meta.args } : {}),
        ...(meta.saveReport === true ? { saveReport: true } : {}),
        ...(meta.onPending ? { onPending: meta.onPending } : {}),
        ...(meta.onFailure ? { onFailure: meta.onFailure } : {}),
      },
    ]
  })
}

function parseItemsFromAnyRaw(raw: string): TodoItemState[] {
  const json = parseJsonRecord(raw)
  return json ? normalizeItems(json.items) : []
}

function normalizeItems(value: unknown): TodoItemState[] {
  return Array.isArray(value) ? (value.filter((item) => item && typeof item === "object") as TodoItemState[]) : []
}

function parseJsonRecord(raw: string): Record<string, unknown> | null {
  try {
    const parsed = JSON.parse(raw) as unknown
    return recordField(parsed)
  } catch {
    return null
  }
}

function recordField(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {}
}

function stringField(value: unknown): string {
  return typeof value === "string" ? value : ""
}

function stringArray(value: unknown): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : []
}

File: src/runtime-services/goal-scheduler/managed_todo_state.py
"""Managed-goal todo state contract for goal-scheduler.

The scheduler owns timing and dispatch. This module owns the todo JSON shape
used for managed goal state.
"""

import json


def is_managed_todo_text(text: str) -> bool:
    data = parse_json_object(text)
    return is_managed_todo_data(data) if data is not None else False


def is_managed_todo_data(data: dict) -> bool:
    return (
        data.get("managed") is True
        or data.get("managed") == "true"
        or data.get("managedModel") in ("agentGoal", "agentLoop")
    )


def parse_json_object(text: str) -> dict | None:
    try:
        data = json.loads(text)
    except Exception:
        return None
    return data if isinstance(data, dict) else None


def string_list(value: object) -> list[str]:
    return [item for item in value if isinstance(item, str)] if isinstance(value, list) else []


def route_from_items(items: list[dict]) -> list[dict]:
    route: list[dict] = []
    for item in items:
        meta = item.get("meta") if isinstance(item.get("meta"), dict) else {}
        stage = meta.get("stage")
        evidence = meta.get("evidence") or item.get("id")
        capability = meta.get("capability")
        if isinstance(stage, str) and isinstance(evidence, str) and isinstance(capability, str):
            step = {"stage": stage, "evidence": evidence, "capability": capability}
            if isinstance(meta.get("args"), dict):
                step["args"] = meta["args"]
            if meta.get("saveReport") is True:
                step["saveReport"] = True
            route.append(step)
    return route


def parse_json_todo_goal_state(goal_id: str, data: dict) -> dict:
    items = [item for item in data.get("items", []) if isinstance(item, dict)] if isinstance(data.get("items"), list) else []
    raw_destination = data.get("destination") if isinstance(data.get("destination"), dict) else {}
    route = data.get("route") if isinstance(data.get("route"), list) else route_from_items(items)
    evidence = string_list(raw_destination.get("evidence")) or string_list(data.get("evidence"))
    if not evidence:
        evidence = [
            str((item.get("meta") if isinstance(item.get("meta"), dict) else {}).get("evidence") or item.get("id"))
            for item in items
            if item.get("id")
        ]
    facts = data.get("facts") if isinstance(data.get("facts"), dict) else {}
    facts = dict(facts)
    for item in items:
        meta = item.get("meta") if isinstance(item.get("meta"), dict) else {}
        key = meta.get("evidence") or item.get("id")
        if isinstance(key, str):
            facts[key] = item.get("completed") is True
    capabilities = string_list(data.get("capabilities"))
    if not capabilities:
        capabilities = [step["capability"] for step in route if isinstance(step.get("capability"), str)]
    outcome = data.get("description") if isinstance(data.get("description"), str) else raw_destination.get("outcome")
    parsed = dict(data)
    parsed.update(
        {
            "id": goal_id,
            "version": data.get("version", 1),
            "state": data.get("state", "active"),
            "type": data.get("type", "general"),
            "destination": {**raw_destination, "outcome": outcome if isinstance(outcome, str) else "", "evidence": evidence},
            "capabilities": capabilities,
            "route": route,
            "facts": facts,
            "blockers": string_list(data.get("blockers")),
        }
    )
    return parsed


def parse_todo_goal_state(goal_id: str, text: str) -> dict:
    data = parse_json_object(text)
    if data is None:
        raise ValueError(f"goal {goal_id} todo state must be JSON")
    return parse_json_todo_goal_state(goal_id, data)


def todo_items_from_state(data: dict, now: str) -> list[dict]:
    destination = data.get("destination") if isinstance(data.get("destination"), dict) else {}
    evidence = string_list(destination.get("evidence"))
    route = data.get("route") if isinstance(data.get("route"), list) else []
    facts = data.get("facts") if isinstance(data.get("facts"), dict) else {}
    route_by_evidence = {
        step.get("evidence"): step
        for step in route
        if isinstance(step, dict) and isinstance(step.get("evidence"), str)
    }
    if evidence:
        items = []
        for key in evidence:
            step = route_by_evidence.get(key) if isinstance(route_by_evidence.get(key), dict) else {}
            completed = facts.get(key) is True
            items.append(
                {
                    "id": key,
                    "title": step.get("stage") if isinstance(step.get("stage"), str) else key,
                    "body": "",
                    "assignee": None,
                    "completed": completed,
                    "createdAt": data.get("createdAt") if isinstance(data.get("createdAt"), str) else now,
                    "completedAt": data.get("updatedAt") if completed and isinstance(data.get("updatedAt"), str) else None,
                    "meta": {
                        "evidence": key,
                        **({"stage": step["stage"]} if isinstance(step.get("stage"), str) else {}),
                        **({"capability": step["capability"]} if isinstance(step.get("capability"), str) else {}),
                        **({"args": step["args"]} if isinstance(step.get("args"), dict) else {}),
                        **({"saveReport": True} if step.get("saveReport") is True else {}),
                    },
                }
            )
        return items
    return [
        {
            "id": capability,
            "title": capability,
            "body": "",
            "assignee": None,
            "completed": False,
            "createdAt": data.get("createdAt") if isinstance(data.get("createdAt"), str) else now,
            "completedAt": None,
            "meta": {"capability": capability},
        }
        for capability in string_list(data.get("capabilities"))
    ]


def serialize_todo_goal_state(goal_id: str, data: dict, now: str) -> str:
    destination = data.get("destination") if isinstance(data.get("destination"), dict) else {}
    outcome = destination.get("outcome") if isinstance(destination.get("outcome"), str) else ""
    record = dict(data)
    record["version"] = 1
    record["id"] = goal_id
    record["title"] = goal_id
    record["description"] = outcome
    record["createdAt"] = data.get("createdAt") if isinstance(data.get("createdAt"), str) else now
    record["managed"] = True
    record["managedModel"] = "agentLoop" if record.get("scheduleMode") == "agentLoop" or record.get("type") == "agentLoop" else "agentGoal"
    record["evidence"] = string_list(destination.get("evidence"))
    record["items"] = todo_items_from_state(data, now)
    return json.dumps(record, indent=2) + "\n"
